File: mobile/src/lib/parserConfigSync.js
// ─────────────────────────────────────────────────────────────────────────────
// On every app init, try to fetch the latest parser-config.json and
// parser-config-wordmap.json straight from GitHub. Each file independently
// goes through the same three-tier fallback:
//   1. Fetch succeeds → cache it and use it.
//   2. Fetch fails (offline, GitHub down, malformed response) → use the last
//      successfully cached version, if one exists from a previous launch.
//   3. No cache either (e.g. first-ever launch with no network) → use the
//      version bundled at build time.
// They're synced independently (not as one atomic pair) so a hiccup fetching
// one doesn't hold back a fresh copy of the other.
// This is what lets report-format fixes (new keyword/reference-range/column-
// pattern data) reach users without an app store release — parser-core.mjs's
// logic itself never changes here, only the data it's configured with.
// ─────────────────────────────────────────────────────────────────────────────
import { Filesystem, Directory, Encoding } from '@capacitor/filesystem';
import { configureParser } from '../../../parser-core.mjs';
import bundledConfig from '../../../parser-config.json';
import bundledWordMap from '../../../parser-config-wordmap.json';

const GITHUB_BASE = 'https://raw.githubusercontent.com/pocha/rakta-sutra/refs/heads/main';
const FETCH_TIMEOUT_MS = 5000;

function isValidConfig(c) {
  return !!(c && c.keywordMap && c.markerGroups && c.valueLimits && c.refRanges && c.layout);
}

// The word-map has no fixed shape to validate beyond "an object of arrays" —
// unlike the main config it's just KEYWORD -> [marker, ...] pairs throughout.
function isValidWordMap(w) {
  return !!(w && typeof w === 'object' && !Array.isArray(w)
    && Object.values(w).every(v => Array.isArray(v)));
}

// AbortSignal.timeout() is a newer API that may not exist on the older end
// of our iOS 15.0 deployment target — build the timeout by hand with
// AbortController instead, which has much wider WebKit support.
function fetchWithTimeout(url, ms) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(new Error(`timed out after ${ms}ms`)), ms);
  return fetch(url, { signal: controller.signal }).finally(() => clearTimeout(timer));
}

// Runs the fetch → cache → bundled fallback chain for one file. Returns
// { data, source } rather than throwing — sync failures for one file should
// never prevent the other from loading.
async function syncOne({ label, remoteUrl, cachePath, bundled, isValid }) {
  try {
    const res = await fetchWithTimeout(remoteUrl, FETCH_TIMEOUT_MS);
    if (!res.ok) throw new Error(`fetch failed: ${res.status}`);
    const remote = await res.json();
    if (!isValid(remote)) throw new Error('fetched data failed shape validation');

    await Filesystem.writeFile({
      path: cachePath,
      directory: Directory.Data,
      data: JSON.stringify(remote),
      encoding: Encoding.UTF8,
    });
    return { data: remote, source: 'remote' };
  } catch (err) {
    console.error(`[parserConfig] ${label}: fetch failed, falling back to cache:`, err);
  }

  try {
    const { data } = await Filesystem.readFile({ path: cachePath, directory: Directory.Data, encoding: Encoding.UTF8 });
    const cached = JSON.parse(data);
    if (!isValid(cached)) throw new Error('cached data failed shape validation');
    return { data: cached, source: 'cache' };
  } catch (err) {
    console.error(`[parserConfig] ${label}: no usable cache, falling back to bundled:`, err);
    return { data: bundled, source: 'bundled' };
  }
}

export async function initParserConfig() {
  const [config, wordMap] = await Promise.all([
    syncOne({
      label: 'config',
      remoteUrl: `${GITHUB_BASE}/parser-config.json`,
      cachePath: 'parser-config.json',
      bundled: bundledConfig,
      isValid: isValidConfig,
    }),
    syncOne({
      label: 'wordmap',
      remoteUrl: `${GITHUB_BASE}/parser-config-wordmap.json`,
      cachePath: 'parser-config-wordmap.json',
      bundled: bundledWordMap,
      isValid: isValidWordMap,
    }),
  ]);

  configureParser(config.data, wordMap.data);
  console.log(`[parserConfig] using config (${config.source}) + wordmap (${wordMap.source})`);
}

File: mobile_flutter/lib/services/parser_config_sync.dart
// Port of mobile/src/lib/parserConfigSync.js. On every app init, tries to
// fetch the latest parser-config.json and parser-config-wordmap.json
// straight from GitHub. Each file independently goes through the same
// three-tier fallback:
//   1. Fetch succeeds → cache it and use it.
//   2. Fetch fails (offline, GitHub down, malformed response) → use the
//      last successfully cached version, if one exists from a previous launch.
//   3. No cache either (e.g. first-ever launch with no network) → use the
//      version bundled at build time.
// They're synced independently (not as one atomic pair) so a hiccup fetching
// one doesn't hold back a fresh copy of the other. This is what lets
// report-format fixes (new keyword/reference-range/column-pattern data)
// reach users without an app store release — parser-core.mjs's logic
// itself never changes here, only the data it's configured with.
import 'dart:convert';
import 'dart:io';
import 'package:crypto/crypto.dart';
import 'package:flutter/services.dart' show rootBundle;
import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'db.dart';
import 'parser_bridge.dart';
import 'parser_config.dart';
import 'reparse_all.dart';

const _configHashKey = 'lastConfigHash';
const _githubBase = 'https://raw.githubusercontent.com/pocha/rakta-sutra/refs/heads/main';
const _fetchTimeout = Duration(seconds: 5);

bool _isValidConfig(Map<String, dynamic> c) =>
    c['keywordMap'] != null && c['markerGroups'] != null && c['valueLimits'] != null && c['refRanges'] != null && c['units'] != null && c['layout'] != null;

// The word-map has no fixed shape to validate beyond "an object of arrays" —
// unlike the main config it's just KEYWORD -> [marker, ...] pairs throughout.
bool _isValidWordMap(Map<String, dynamic> w) => w.values.every((v) => v is List);

class _SyncResult {
  final Map<String, dynamic> data;
  final String source; // 'remote' | 'cache' | 'bundled'
  _SyncResult(this.data, this.source);
}

Future<_SyncResult> _syncOne({
  required String label,
  required String remoteUrl,
  required String cacheFileName,
  required String bundledAssetPath,
  required bool Function(Map<String, dynamic>) isValid,
}) async {
  final dir = await getApplicationSupportDirectory();
  final cacheFile = File(p.join(dir.path, cacheFileName));

  try {
    final res = await http.get(Uri.parse(remoteUrl)).timeout(_fetchTimeout);
    if (res.statusCode != 200) throw Exception('fetch failed: ${res.statusCode}');
    final remote = jsonDecode(res.body) as Map<String, dynamic>;
    if (!isValid(remote)) throw Exception('fetched data failed shape validation');
    await cacheFile.writeAsString(res.body);
    return _SyncResult(remote, 'remote');
  } catch (err) {
    // ignore: avoid_print
    print('[parserConfig] $label: fetch failed, falling back to cache: $err');
  }

  try {
    final cached = jsonDecode(await cacheFile.readAsString()) as Map<String, dynamic>;
    if (!isValid(cached)) throw Exception('cached data failed shape validation');
    return _SyncResult(cached, 'cache');
  } catch (err) {
    // ignore: avoid_print
    print('[parserConfig] $label: no usable cache, falling back to bundled: $err');
    final bundled = jsonDecode(await rootBundle.loadString(bundledAssetPath)) as Map<String, dynamic>;
    return _SyncResult(bundled, 'bundled');
  }
}

String _hashConfig(Map<String, dynamic> config, Map<String, dynamic> wordMap) =>
    sha256.convert(utf8.encode(jsonEncode(config) + jsonEncode(wordMap))).toString();

// Fire-and-forget from initParserConfig() — must never block app startup on
// hashing (cheap) or a potentially-slow reparse of every stored report.
//
// Reparses whenever the recorded hash differs, including the very first
// time this check ever runs on a device — an install can already have
// reports sitting in its DB from before this tracking existed, so "no hash
// recorded yet" doesn't mean "nothing's stale".
Future<void> _reparseIfConfigChanged(Map<String, dynamic> config, Map<String, dynamic> wordMap) async {
  final hash = _hashConfig(config, wordMap);
  final previous = await Db.instance.getDeviceSetting(_configHashKey);
  if (previous != hash) {
    // ignore: avoid_print
    print('[parserConfig] config hash $previous -> $hash — reparsing all stored reports');
    await reparseAllReports();
    await Db.instance.setDeviceSetting(_configHashKey, hash);
  }
}

Future<void> initParserConfig() async {
  final results = await Future.wait([
    _syncOne(
      label: 'config',
      remoteUrl: '$_githubBase/parser-config.json',
      cacheFileName: 'parser-config.json',
      bundledAssetPath: 'assets/parser-config.json',
      isValid: _isValidConfig,
    ),
    _syncOne(
      label: 'wordmap',
      remoteUrl: '$_githubBase/parser-config-wordmap.json',
      cacheFileName: 'parser-config-wordmap.json',
      bundledAssetPath: 'assets/parser-config-wordmap.json',
      isValid: _isValidWordMap,
    ),
  ]);
  final config = results[0];
  final wordMap = results[1];

  ParserConfig.reconfigure(config.data, wordMap.data);
  await ParserBridge.instance.configureParser(config.data, wordMap.data);
  // ignore: avoid_print
  print('[parserConfig] using config (${config.source}) + wordmap (${wordMap.source})');

  _reparseIfConfigChanged(config.data, wordMap.data).catchError((err) {
    // ignore: avoid_print
    print('[parserConfig] reparse-on-config-change check failed: $err');
  });
}
